bumble/profiles/hap.py

```diff
@@ -0,0 +1,590 @@
+    def __init__(self, index):
+        self.change_id = PresetChangedOperation.ChangeId.PRESET_RECORD_AVAILABLE
+        self.additional_parameters = index
+
+
+class PresetChangedOperationUnavailable(PresetChangedOperation):
+    def __init__(self, index):
+        self.change_id = PresetChangedOperation.ChangeId.PRESET_RECORD_UNAVAILABLE
+        self.additional_parameters = index
+
+
+class PresetRecord:
+    '''See Hearing Access Service 2.8. Preset record.'''
+
+    class Property:
+        class Writable(OpenIntEnum):
+            CANNOT_BE_WRITTEN = 0b0
+            CAN_BE_WRITTEN = 0b1
+
+        class IsAvailable(OpenIntEnum):
+            IS_UNAVAILABLE = 0b0
+            IS_AVAILABLE = 0b1
+
+        writable: Writable
+        is_available: IsAvailable
+
+        def __bytes__(self) -> bytes:
+            return bytes([self.writable | (self.is_available << 1)])
+
+    index: int
+    properties: Property
+    name: str
+
+    def __bytes__(self) -> bytes:
+        return bytes([self.index]) + bytes(self.properties) + self.name.encode('utf-8')
+
+    def is_available(self) -> bool:
+        return (
+            self.properties.is_available
+            == PresetRecord.Property.IsAvailable.IS_AVAILABLE
+        )
+
+
+# -----------------------------------------------------------------------------
+# Server
+# -----------------------------------------------------------------------------
+class HearingAccessService(gatt.TemplateService):
+    UUID = gatt.GATT_HEARING_ACCESS_SERVICE
+
+    hearing_aid_features_characteristic: gatt.Characteristic
+    hearing_aid_preset_control_point: gatt.Characteristic
+    active_preset_index_characteristic: gatt.Characteristic
+    active_preset_index: int = 0x00
+    active_preset_index_per_device: Dict[
+        Address, int
+    ]  # TODO validate address type → public_address ?
+
+    device: Device
+
+    hearing_aid_features: HearingAidFeatures
+    preset_records: Dict[int, PresetRecord] = {}  # key is the preset index
+    read_presets_request_in_progress: bool = False
+
+    preset_changed_operations_history: List[PresetChangedOperation]
+    preset_changed_operations_per_device: Dict[Address, int]
+    preset_changed_operations_history_per_device: Dict[
+        Address, List[PresetChangedOperation]
+    ] = {}
+
+    # Keep an updated list of connected client to send notification to
+    currently_connected_clients: Set[Connection]
+
+    def __init__(
+        self, device: Device, features: HearingAidFeatures, presets: List[PresetRecord]
+    ) -> None:
+        self.device = device
+        self.hearing_aid_features = features
+        for p in presets:
+            self.preset_records[p.index] = p
+
+        with contextlib.closing(EventWatcher()) as watcher:
+
+            @watcher.on(
+                device, 'connection'
+            )  # should it be on_characteristic_subscription ?
+            async def on_connection(connection: Connection) -> None:
```

Comment:
Currently we don't introduce `AsyncIOEventEmitter` so I guess this won't work?

Reply:
You mean the method the watcher call cannot be a async method ?

Reply:
I provide a way in the suggestion.